tests/test_matrix.py

```diff
@@ -0,0 +1,271 @@
+                         list(zip(ILLEGAL_VALUE_MATRIX_LIST, [ValueError] * len(ILLEGAL_VALUE_MATRIX_LIST))))
+def test_is_valid_matrix(test_input, expected_exception):
+    if expected_exception is None:
+        Matrix.is_valid_matrix(test_input)
+    else:
+        with pytest.raises(expected_exception):
+            Matrix.is_valid_matrix(test_input)
+
+
+########################################################################################################################
+@pytest.mark.parametrize('expected_matrix', LEGAL_MATRIX_LIST)
+def test_matrix_tuples(matrix_factory, expected_matrix):
+    assert matrix_factory(expected_matrix).tuples == expected_matrix
+
+
+########################################################################################################################
+@pytest.mark.parametrize('expected_matrix', LEGAL_MATRIX_LIST)
+def test_matrix_str(matrix_factory, expected_matrix):
+    assert str(matrix_factory(expected_matrix)) == f"{expected_matrix}"
+
+
+########################################################################################################################
+@pytest.mark.parametrize('expected_matrix', LEGAL_MATRIX_LIST)
+def test_matrix_repr(matrix_factory, expected_matrix):
+    assert repr(matrix_factory(expected_matrix)) == f"Matrix({expected_matrix})"
+
+
+########################################################################################################################
+@pytest.mark.parametrize('expected_matrix', LEGAL_MATRIX_LIST)
+def test_matrix_len(matrix_factory, expected_matrix):
+    assert len(matrix_factory(expected_matrix)) == len(expected_matrix)
+
+
+########################################################################################################################
+@pytest.mark.parametrize('expected_matrix', LEGAL_MATRIX_LIST)
+def test_matrix_getitem(compare_matrix_rows, matrix_factory, expected_matrix):
+    compare_matrix_rows(matrix_factory(expected_matrix), expected_matrix)
+
+
+########################################################################################################################
+@pytest.mark.parametrize('expected_matrix', LEGAL_MATRIX_LIST)
+def test_matrix_hash(matrix_factory, expected_matrix):
+    assert hash(matrix_factory(expected_matrix)) == hash(expected_matrix)
+
+
+########################################################################################################################
+@pytest.mark.parametrize('expected_matrix, expected_result', zip(
+    LEGAL_MATRIX_LIST,
+    [(()), ((-1,),), ((-1, -2), (3, 4)), ((-1, -2.5, -3), (-3, -4, -5), (-1, 5.5, -6))])
+)
+def test_matrix_neg(matrix_factory, expected_matrix, expected_result):
+    assert (-matrix_factory(expected_matrix)).matrix == expected_result
+
+
+########################################################################################################################
+@pytest.mark.parametrize('expected_matrix', LEGAL_MATRIX_LIST)
```

Comment:
For the expected, make a list of actual Matrix instead of list of tuples you turn in to matrix